codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/customizations/TsServerCodegenDecorator.kt

```diff
@@ -0,0 +1,150 @@
+    override val order: Byte = 0
+    private val napi_build = TsServerCargoDependency.NapiBuild.toType()
+
+    override fun extras(codegenContext: ServerCodegenContext, rustCrate: RustCrate) {
+        rustCrate.withFile("build.rs") {
+            rustTemplate(
+                """
+                fn main() {
+                    #{napi_build}::setup();
+                }
+                """,
+                "napi_build" to napi_build,
+            )
+        }
+    }
+}
+
+class NapiPackageJsonDecorator : ServerCodegenDecorator {
+    override val name: String = "NapiPackageJsonDecorator"
+    override val order: Byte = 0
+
+    override fun extras(codegenContext: ServerCodegenContext, rustCrate: RustCrate) {
+        val name = codegenContext.settings.moduleName.toSnakeCase()
+        val version = codegenContext.settings.moduleVersion
+
+        // TODO(https://github.com/awslabs/smithy-rs/issues/2317): we should probably use a real JSON writer, but I did not want to add
+        //  other external libraries at this stage.
+        rustCrate.withFile("package.json") {
+            val content = """{
+                "name": "@amzn/$name",
+                "version": "$version",
+                "main": "index.js",
+                "types": "index.d.ts",
+                "napi": {
+                    "name": "$name",
+                    "triple": {}
+                },
+                "devDependencies": {
+                    "@napi-rs/cli": ">=2",
+                    "@types/node": ">=18"
+                },
+                "engines": {
+                    "node": ">=10"
+                },
+                "scripts": {
+                    "artifacts": "napi artifacts",
+                    "build": "napi build --platform --release",
+                    "build:debug": "napi build --platform",
+                    "prepublishOnly": "napi prepublish -t npm",
+                    "universal": "napi universal",
+                    "version": "napi version"
+                },
+                "packageManager": "yarn",
+                "dependencies": {
+                    "yarn": ">=1"
+                }
+}"""
+            this.write(content)
+        }
+        val operations = codegenContext.serviceShape.operations
+        val model = codegenContext.model
+        val imports =
+            operations.joinToString {
+                listOf(
+                    model.expectShape(it).asOperationShape().get().inputShape.name,
+                    model.expectShape(it).asOperationShape().get().outputShape.name,
+                ).joinToString()
+            }
+        val handlers =
+            operations.joinToString("\n") {
+                "async ${it.name}(_: ${
+                model.expectShape(it).asOperationShape().get().inputShape.name
+                }): Promise<${model.expectShape(it).asOperationShape().get().outputShape.name}> { /* TODO */ }"
+            }
+        rustCrate.withFile("app.ts") {
```

Comment:
@pose this should go away, we will make this documentation based in the future.

Reply:
Done (resolve button doesn't work yet).